Student ID: FIN2104319</br>
Student Name: NG TEK YAW

* This page is designed solely for showing the steps in calculating weekly's capital allocation by using Kelly Formula.

### <center><u>Kelly Formula</u></center>
$$ K\% = W - \frac{(1 - W)}{R} $$
<u>where</u></br>
W: Winning Probability</br>
R: Win/Loss Ratio

In [1]:
import pandas as pd
import numpy as np 
from pandas_datareader import data as web

### <font color='lightgrey'>First, we will define a function to calculate Kelly Formula</font>

In [2]:
def kelly(w: float, r: float) -> float:
    """
    w: winning probability
    r: win/loss ratio
    """
    return w - ((1 - w) / r)

# Calculation Part

### Week 1's Capital Allocation

In [3]:
## week 1's portfolio
## there are 12 stocks included in the week1's portfolio
tickers = [
    "CTRA",
    "MRO",
    "OXY",
    "CLW",
    "MERC",
    "IP",
    "SEOAY",
    "DAL",
    "HSY",
    "PM",
    "MO",
]
tickers

['CTRA', 'MRO', 'OXY', 'CLW', 'MERC', 'IP', 'SEOAY', 'DAL', 'HSY', 'PM', 'MO']

#### Scraping historical price of the stocks in the portfolio

In [4]:
df1 = pd.DataFrame(columns=tickers)
for ticker in tickers:
    df1[ticker] = web.DataReader(ticker, 'yahoo', '1980', '2022-10-17')['Adj Close']
df1

CTRA        MRO        OXY        CLW       MERC         IP  \
Date                                                                           
1990-02-08   1.001873   3.552361   4.045230        NaN   6.569688   8.378754   
1990-02-09   0.994613   3.590421   4.101156        NaN   6.569688   8.463390   
1990-02-12   0.987353   3.552361   4.101156        NaN   6.660933   8.315283   
1990-02-13   0.980093   3.565047   4.138438        NaN   6.569688   8.336440   
1990-02-14   0.972833   3.565047   4.082514        NaN   6.752179   8.336440   
...               ...        ...        ...        ...        ...        ...   
2022-10-11  28.103802  26.965120  64.029999  38.360001  13.830000  31.180830   
2022-10-12  29.079630  27.364014  65.110001  38.209999  14.030000  31.101841   
2022-10-13  29.723675  28.251547  68.000000  39.080002  14.380000  32.365662   
2022-10-14  27.732988  26.975092  66.680000  38.450001  14.330000  32.207684   
2022-10-17  28.220903  27.064842  67.360001  39.540001  14.300000  32.809975   

            SEOAY        DAL         HSY         PM         MO  
Date                                                            
1990-02-08    NaN        NaN    3.769056        NaN   0.567445  
1990-02-09    NaN        NaN    3.798968        NaN   0.565643  
1990-02-12    NaN        NaN    3.709228        NaN   0.556636  
1990-02-13    NaN        NaN    3.709228        NaN   0.558438  
1990-02-14    NaN        NaN    3.679315        NaN   0.563842  
...           ...        ...         ...        ...        ...  
2022-10-11  13.06  28.840000  224.413284  85.620003  44.349998  
2022-10-12  13.05  29.209999  223.487595  85.669998  44.750000  
2022-10-13  13.20  30.379999  226.742416  87.230003  45.889999  
2022-10-14  13.34  31.080000  221.676056  85.260002  45.230000  
2022-10-17  13.93  31.129999  224.672073  85.570000  44.590000  

[8236 rows x 11 columns]

* We will then calculate the portfolio return(equally-weighted portfolio return), and then treating the whole portfolio as a single asset. 

In [46]:

df_return = (df1
                .assign(** {"portfolio return" : lambda x: x.sum(axis=1) / 12})
                .resample('W')
                .agg(lambda x: (x[-1] - x[0]) / x[0] * 100)
)
df_return_ = df_return.iloc[-100:]  ### we are accessing to the last 100 trades performance

In [6]:
df_return_

CTRA        MRO        OXY       CLW       MERC        IP  \
Date                                                                        
2020-11-29  0.794117   2.917327   5.076135 -1.637487  -1.904762  1.359724   
2020-12-06 -6.164382  15.709470  17.956853  4.755084  16.971933 -1.232807   
2020-12-13  2.843309   5.530641  14.903371  2.601003   4.926634 -0.082189   
2020-12-20 -1.623189   0.148813  -3.390703 -5.058263  -0.976568  4.103306   
2020-12-27  0.474486   1.221382   1.144814  8.410405   3.106897  0.565893   
...              ...        ...        ...       ...        ...       ...   
2022-09-25 -8.914993 -15.541846  -9.199135 -5.724820 -10.494752 -9.364738   
2022-10-02  3.159563   7.064961   6.167933  0.534751   5.263164 -0.220332   
2022-10-09  6.725146  11.894279   8.960359 -2.203435   5.034861 -3.011679   
2022-10-16 -0.175625  -1.707849   1.599883 -0.953114   4.522246  3.325936   
2022-10-23  0.000000   0.000000   0.000000  0.000000   0.000000  0.000000   

                SEOAY        DAL       HSY        PM        MO  \
Date                                                             
2020-11-29   5.446755   5.852025  0.830906 -0.952605  0.422863   
2020-12-06   0.775644   5.242238  2.015010  6.600659  3.866404   
2020-12-13   5.329486  -1.440370 -0.407439  4.064622  3.142251   
2020-12-20   7.142856  -0.024572  1.166705  2.002356  2.051284   
2020-12-27  -0.536467  -1.021425  0.193782 -0.554401 -0.852201   
...               ...        ...       ...       ...       ...   
2022-09-25 -12.405239 -11.900420  0.480108 -4.305663 -2.866464   
2022-10-02   0.877898   0.142752 -1.479131 -6.625413 -2.628406   
2022-10-09  -3.619913   3.998582 -1.799072 -0.446898  3.504957   
2022-10-16   2.852736   5.642420 -0.375750  1.500003  4.072712   
2022-10-23   0.000000   0.000000  0.000000  0.000000  0.000000   

            portfolio return  
Date                          
2020-11-29          1.115809  
2020-12-06          3.817119  
2020-12-13          1.956127  
2020-12-20          0.899923  
2020-12-27          0.642017  
...                      ...  
2022-09-25         -4.799089  
2022-10-02         -0.613120  
2022-10-09          1.187223  
2022-10-16          1.012391  
2022-10-23          0.000000  

[100 rows x 12 columns]

In [48]:
win_loss = (df_return_
        .loc[:, 'portfolio return']
        .apply(lambda x: 'win' if x > 0 else "loss")
        .value_counts()
)
win_loss  ### from the past 100 trades, it is proven that out of 1707 week, the portfolio provide 925 times of positive return
###  and 782 times of losses. 


win     61
loss    39
Name: portfolio return, dtype: int64

In [53]:
W = win_loss['win'] / (win_loss.sum())
B = df_return_.query("`portfolio return` > 0").mean().iloc[-1]
A = np.abs(df_return_.query("`portfolio return` < 0").mean().iloc[-1])
R = B / A
print(f"W: {W}\nR: {B/A:.3f}")
float(kelly(W, R)) * 100

W: 0.62
R: 1.048


25.73149667148668

<strong>As a result, the capital that should be allocated to the portfolio(assumed to be a single asset) will be roughly <font color='red'>26%</font></strong>

### Week 2's Capital Allocation

In [10]:
tickers = [
    "FANG",
    "CTRA",
    "MRO",
    "JBLU",
    "DAL",
    "AAPL",
    "ADBE",
    "pfe",
    "VRTX",
    "MCK",
    "qdel",
    "KO",
]

In [13]:
df2 = pd.DataFrame(columns=tickers)
for ticker in tickers:
    df2[ticker] = web.DataReader(ticker, 'yahoo', '1980', '2022-10-24')['Adj Close']
df2

FANG       CTRA        MRO  JBLU        DAL        AAPL  \
Date                                                                        
2012-10-12   15.341786  17.712156  24.704889  5.09   8.900230   19.252886   
2012-10-15   15.560956  17.885803  25.116917  5.19   8.980172   19.407286   
2012-10-16   15.596024  18.414825  25.764385  5.20   8.873587   19.866814   
2012-10-17   15.517124  18.479435  26.344593  5.26   9.051234   19.708443   
2012-10-18   15.517124  18.192720  26.260508  5.31   9.068999   19.342464   
...                ...        ...        ...   ...        ...         ...   
2022-10-18  139.930710  28.728333  27.483679  6.98  32.169998  143.511932   
2022-10-19  144.054138  29.391895  28.271492  6.85  32.520000  143.621750   
2022-10-20  144.606567  28.972290  28.151825  7.01  32.150002  143.152527   
2022-10-21  150.259048  28.952772  29.069277  7.16  32.580002  147.026108   
2022-10-24  151.186310  29.548027  29.388390  7.54  33.270000  149.202484   

                  ADBE        pfe        VRTX         MCK       qdel  \
Date                                                                   
2012-10-12   31.940001  16.514114   53.529999   82.326416  17.100000   
2012-10-15   32.330002  16.809952   53.240002   83.801781  17.280001   
2012-10-16   33.240002  16.961157   53.290001   85.249496  17.469999   
2012-10-17   33.439999  16.987446   53.950001   84.152176  17.590000   
2012-10-18   33.470001  17.118927   52.630001   84.225937  17.500000   
...                ...        ...         ...         ...        ...   
2022-10-18  292.980011  43.715324  300.170013  360.880005  81.339996   
2022-10-19  299.829987  42.743652  292.019989  360.660004  77.889999   
2022-10-20  302.380005  42.545353  290.160004  359.190002  78.720001   
2022-10-21  306.369995  44.568016  300.000000  370.640015  78.750000   
2022-10-24  316.220001  45.153004  310.500000  379.109985  79.669998   

                   KO  
Date                   
2012-10-12  27.915070  
2012-10-15  27.842060  
2012-10-16  27.674112  
2012-10-17  27.557276  
2012-10-18  27.630312  
...               ...  
2022-10-18  56.439999  
2022-10-19  55.959999  
2022-10-20  55.080002  
2022-10-21  55.959999  
2022-10-24  57.570000  

[2524 rows x 12 columns]

In [111]:
df_return_2 = (df2
                .assign(** {"portfolio return" : lambda x: x.sum(axis=1) / 12})
                .resample('W')
                .agg(lambda x: (x[-1] - x[0]) / x[0] * 100)
)
df_return_2_ = df_return_2.iloc[-120:]  ### we are accessing to the last 100 trades performance

In [114]:
win_loss = (df_return_2_
        .loc[:, 'portfolio return']
        .apply(lambda x: 'win' if x > 0 else "loss")
        .value_counts()
)
win_loss

win     81
loss    39
Name: portfolio return, dtype: int64

In [115]:
W = win_loss['win'] / (win_loss.sum())
B = df_return_2_.query("`portfolio return` > 0").mean().iloc[-1]
A = np.abs(df_return_2_.query("`portfolio return` < 0").mean().iloc[-1])
R = B / A
print(f"W: {W}\nR: {B/A:.3f}")
float(kelly(W, R)) * 100

W: 0.675
R: 0.923


32.29069931781551

* The capital allocated to overall portfolio will be 32.29 * 2 = 64%. (Here we multiply kelly by two because we expect the market in week 2 will be bullish)

### Week 3's Capital Allocation

In [118]:
tickers = [
    "XOM",
    "VRTX",
    "AAPL",
    "JBLU",
    "MCK",
    "QDEL",
    "HLT",
    "DAL",
    "ADBE",
    "CGNX",
    "MEDP",
    "RC",
]

In [120]:
df3 = pd.DataFrame(columns=tickers)
for ticker in tickers:
    df3[ticker] = web.DataReader(ticker, 'yahoo', '1980', '2022-10-24')['Adj Close']
df3

XOM        VRTX        AAPL  JBLU         MCK       QDEL  \
Date                                                                          
1979-12-31    0.544875         NaN         NaN   NaN         NaN        NaN   
1980-01-02    0.532520         NaN         NaN   NaN         NaN        NaN   
1980-01-03    0.518928         NaN         NaN   NaN         NaN        NaN   
1980-01-04    0.523871         NaN         NaN   NaN         NaN        NaN   
1980-01-07    0.521400         NaN         NaN   NaN         NaN        NaN   
...                ...         ...         ...   ...         ...        ...   
2022-10-18   99.995018  300.170013  143.511932  6.98  360.880005  81.339996   
2022-10-19  102.961136  292.019989  143.621750  6.85  360.660004  77.889999   
2022-10-20  103.100021  290.160004  143.152527  7.01  359.190002  78.720001   
2022-10-21  105.014610  300.000000  147.026108  7.16  370.640015  78.750000   
2022-10-24  105.748695  310.500000  149.202484  7.54  379.109985  79.669998   

                   HLT        DAL        ADBE       CGNX        MEDP     RC  
Date                                                                         
1979-12-31         NaN        NaN         NaN        NaN         NaN    NaN  
1980-01-02         NaN        NaN         NaN        NaN         NaN    NaN  
1980-01-03         NaN        NaN         NaN        NaN         NaN    NaN  
1980-01-04         NaN        NaN         NaN        NaN         NaN    NaN  
1980-01-07         NaN        NaN         NaN        NaN         NaN    NaN  
...                ...        ...         ...        ...         ...    ...  
2022-10-18  128.452560  32.169998  292.980011  44.218216  161.419998  10.97  
2022-10-19  128.552429  32.520000  299.829987  43.389374  157.089996  10.91  
2022-10-20  126.404907  32.150002  302.380005  42.410744  156.630005  10.61  
2022-10-21  131.079529  32.580002  306.369995  43.639027  157.139999  10.87  
2022-10-24  130.480225  33.270000  316.220001  43.549152  158.639999  11.02  

[10797 rows x 12 columns]

In [135]:
df_return_3 = (df3
                .assign(** {"portfolio return" : lambda x: x.sum(axis=1) / 12})
                .resample('W')
                .agg(lambda x: (x[-1] - x[0]) / x[0] * 100)
)
df_return_3_ = df_return_3.iloc[-100:]  ### we are accessing to the last 100 trades performance

win_loss = (df_return_3_
        .loc[:, 'portfolio return']
        .apply(lambda x: 'win' if x > 0 else "loss")
        .value_counts()
)


W = win_loss['win'] / (win_loss.sum())
B = df_return_3_.query("`portfolio return` > 0").mean().iloc[-1]
A = np.abs(df_return_3_.query("`portfolio return` < 0").mean().iloc[-1])
R = B / A
print(f"W: {W}\nR: {B/A:.3f}")
float(kelly(W, R)) * 100 * 2

W: 0.7
R: 0.948


76.69888040465297

* The capital allocated to overall portfolio will be 77%

### Week 4's Capital Allocation

* In week 4, we invest 60% of total investment funds without using Kelly Formula, this is because Kelly Formula is only suitable for allocating funds by taking into accounts of a stock's winning probability. But for short selling, we want to pick the stocks with higher probability to drop in its values.

### Week 5's Capital Allocation

In [141]:
tickers = [
    "xom",
    "Sedg",
    "ADBE",
    "NFLX",
    "JBLU",
    "AAPL",
    "MAR",
    "CHH",
    "DAL",
    "UCL",
    "BA",
    "GOOGL",
]

In [142]:
df4 = pd.DataFrame(columns=tickers)
for ticker in tickers:
    df4[ticker] = web.DataReader(ticker, 'yahoo', '1980', '2022-10-24')['Adj Close']
df4

xom        Sedg        ADBE        NFLX  JBLU        AAPL  \
Date                                                                           
1979-12-31    0.544875         NaN         NaN         NaN   NaN         NaN   
1980-01-02    0.532519         NaN         NaN         NaN   NaN         NaN   
1980-01-03    0.518929         NaN         NaN         NaN   NaN         NaN   
1980-01-04    0.523871         NaN         NaN         NaN   NaN         NaN   
1980-01-07    0.521400         NaN         NaN         NaN   NaN         NaN   
...                ...         ...         ...         ...   ...         ...   
2022-10-18   99.995018  207.419998  292.980011  240.860001  6.98  143.511932   
2022-10-19  102.961136  199.460007  299.829987  272.380005  6.85  143.621750   
2022-10-20  103.100021  197.839996  302.380005  268.160004  7.01  143.152527   
2022-10-21  105.014610  202.660004  306.369995  289.570007  7.16  147.026108   
2022-10-24  105.748695  203.740005  316.220001  282.450012  7.54  149.202484   

                   MAR         CHH        DAL    UCL          BA       GOOGL  
Date                                                                          
1979-12-31         NaN         NaN        NaN    NaN    1.925760         NaN  
1980-01-02         NaN         NaN        NaN    NaN    1.887719         NaN  
1980-01-03         NaN         NaN        NaN    NaN    1.911494         NaN  
1980-01-04         NaN         NaN        NaN    NaN    2.096939         NaN  
1980-01-07         NaN         NaN        NaN    NaN    2.173017         NaN  
...                ...         ...        ...    ...         ...         ...  
2022-10-18  149.616272  123.910004  32.169998  0.561  137.339996  100.769997  
2022-10-19  149.616272  123.430000  32.520000  0.586  138.389999   99.629997  
2022-10-20  147.152420  119.500000  32.150002  0.597  139.130005   99.970001  
2022-10-21  153.107544  123.599998  32.580002  0.690  141.320007  101.129997  
2022-10-24  154.593826  123.599998  33.270000  0.807  142.050003  102.519997  

[10797 rows x 12 columns]

In [148]:
df_return_4 = (df4
                .assign(** {"portfolio return" : lambda x: x.sum(axis=1) / 12})
                .resample('W')
                .agg(lambda x: (x[-1] - x[0]) / x[0] * 100)
)
df_return_4_ = df_return_4.iloc[-100:]  ### we are accessing to the last 100 trades performance

win_loss = (df_return_4_
        .loc[:, 'portfolio return']
        .apply(lambda x: 'win' if x > 0 else "loss")
        .value_counts()
)


W = win_loss['win'] / (win_loss.sum())
B = df_return_4_.query("`portfolio return` > 0").mean().iloc[-1]
A = np.abs(df_return_4_.query("`portfolio return` < 0").mean().iloc[-1])
R = B / A
print(f"W: {W}\nR: {B/A:.3f}")
float(kelly(W, R)) * 100 * 2

W: 0.67
R: 0.788


50.28915334311853

* The capital allocated to overall portfolio will be 50%